ts/session/medium_group/index.ts

```diff
@@ -39,56 +59,844 @@ async function createSenderKeysForMembers(
+          keyIdx,
+          pubKey: pubKeyBin,
+        };
+      }
+    })
+  );
+
+  return maybeKeys.filter(d => d !== null).map(d => d as RatchetState);
+}
+
+// Create all sender keys based on the changes in
+// the group's composition
+async function getOrCreateSenderKeysForUpdate(
+  groupId: string,
+  members: Array<string>,
+  changes: MemberChanges
+): Promise<SenderKeysContainer> {
+  // 1. Create sender keys for every joining member
+  const joining = changes.joiningMembers || [];
+  const leaving = changes.leavingMembers || [];
+
+  let newKeys = await createSenderKeysForMembers(groupId, joining);
+
+  // 2. Get ratchet states for existing members
+
+  const existingMembers = _.difference(members, joining);
+  // get all devices for members
+
+  const allDevices = _.flatten(
+    await Promise.all(
+      existingMembers.map(m => MultiDeviceProtocol.getAllDevices(m))
+    )
+  );
+
+  let existingKeys: Array<RatchetState> = [];
+
+  if (leaving.length > 0) {
+    // If we have leaving members, we have to re-generate ratchet
+    // keys for existing members
+    const otherKeys = await Promise.all(
+      allDevices.map(async device => {
+        return createSenderKeyForGroup(groupId, PubKey.cast(device));
+      })
+    );
+
+    newKeys = _.union(newKeys, otherKeys);
+  } else {
+    // We can reuse existing keys
+    existingKeys = await getExistingSenderKeysForGroup(groupId, allDevices);
+  }
+
+  return { existingKeys, newKeys };
+}
+
+async function getGroupSecretKey(groupId: string): Promise<Uint8Array> {
+  const groupIdentity = await window.Signal.Data.getIdentityKeyById(groupId);
+  return new Uint8Array(fromHex(groupIdentity.secretKey));
+}
+
+async function syncMediumGroup(group: ConversationModel) {
+  const ourPrimary = await UserUtil.getPrimary();
+
+  const groupId = group.get('id');
+  const members = group.get('members');
+
+  const secretKey = await getGroupSecretKey(groupId);
+
+  const allDevices = _.flatten(
+    await Promise.all(members.map(m => MultiDeviceProtocol.getAllDevices(m)))
+  );
+
+  const secondaryKeys = await MultiDeviceProtocol.getSecondaryDevices(
+    ourPrimary
+  );
+
+  const existingKeys = await getExistingSenderKeysForGroup(groupId, allDevices);
+  const senderKeysForSecondary = await Promise.all(
+    secondaryKeys.map(key => createSenderKeyForGroup(groupId, key))
+  );
+  const senderKeysContainer: SenderKeysContainer = {
+    existingKeys,
+    newKeys: senderKeysForSecondary,
+  };
+
+  const groupUpdate: GroupInfo = {
+    id: group.get('id'),
+    name: group.get('name'),
+    members: group.get('members'),
+    is_medium_group: true,
+    admins: group.get('groupAdmins'),
+    senderKeysContainer,
+    secretKey,
+  };
+
+  // Note: we send this to our primary device which will in effect will send to
+  // our secondary devices, actually ignoring the primary
+  await sendGroupUpdateForMedium(
+    { joiningMembers: [ourPrimary.key] },
+    groupUpdate
+  );
```

Comment:
I think the comment is not correct here.
The device ignored is the one triggering the update (which could be a secondary)

Reply:
but that's not much

Reply:
Updated.